src/services/jsonSchemaService.ts

```diff
@@ -527,17 +585,52 @@ export class JSONSchemaService implements IJSONSchemaService {
 				collectArrayEntries(next.anyOf, next.allOf, next.oneOf, <JSONSchema[]>next.items);
 			};
 
+			const handleId = (next: JSONSchema) => {
+				// TODO figure out should loops be preventse
+				const id = next.$id || next.id;
+				if (typeof id === 'string' && id.charAt(0) === '#') {
+					delete next.$id;
+					delete next.id;
+					// Use a blank separator, as the $id already has the '#'
+					const fullId = reconstructRefURI(parentSchemaURL, id, '');
+
+					if (!resolvedSubSchemas[fullId]) {
+						// This is the place we fill the blanks in
+						resolvedSubSchemas[fullId] = next;
+					} else if (resolvedSubSchemas[fullId] !== next) { 
+						// Duplicate may occur in recursive $refs, but as long as they are the same object
+						// it's ok, otherwise report and error
+						resolveErrors.push(localize('json.schema.duplicateid', 'Duplicate id declaration: \'{0}\'', id));
+					}
+
+					// Resolve all pending requests and cleanup the queue list
+					if (pendingSubSchemas[fullId]) {
+						while (pendingSubSchemas[fullId].length) {
+							const target = pendingSubSchemas[fullId].shift();
+							merge(target!, next);
+						}
+
+						delete pendingSubSchemas[fullId];
+					}
+				}
+			};
+
 			while (toWalk.length) {
 				const next = toWalk.pop()!;
 				if (seen.indexOf(next) >= 0) {
 					continue;
 				}
 				seen.push(next);
+				handleId(next);
```

Comment:
@aeschli I was thinking about the two tests you have added and since I already had a map from the full URI to the full schema, it got me thinking. The root cause of the problem was that the `handleRef` function was called before `handleId`, thus all the schemas with `$ref` and `$id` together were missing that `$ref`. By simply changing the order between them I have managed to solve these two cases without compromising the rest (luckily).